deployment/terraform/create.go

```diff
@@ -340,9 +341,167 @@ func (t *Terraform) PostProcessDatabase(extAgent *ssh.ExtAgent) error {
 		}
 	}
 
+	needsReboot, err := t.HasPendingRebootDBParams()
+	if err != nil {
+		return fmt.Errorf("failed to check whether the DB has pending-reboot parameters")
+	}
+
+	if needsReboot {
+		return t.RebootDBInstances(extAgent)
+	}
+
 	return nil
 }
 
+func (t *Terraform) HasPendingRebootDBParams() (bool, error) {
+	// Build the RDS client
+	cfg, err := t.GetAWSConfig()
+	if err != nil {
+		return false, fmt.Errorf("failed to get AWS config: %w", err)
+	}
+	rdsClient := rds.NewFromConfig(cfg)
+
+	// Check in parallel whether each DB instance needs to be rebooted
+	errChan := make(chan error, len(t.output.DBCluster.Instances))
+	needsRebootChan := make(chan bool, len(t.output.DBCluster.Instances))
+	var wg sync.WaitGroup
+	for _, instance := range t.output.DBCluster.Instances {
+		wg.Add(1)
+		go func(dbId string) {
+			defer wg.Done()
+			needsReboot, err := hasPendingRebootDBParams(rdsClient, dbId)
+			needsRebootChan <- needsReboot
+			errChan <- err
+		}(instance.DBIdentifier)
+	}
+
+	wg.Wait()
+	close(needsRebootChan)
+	close(errChan)
+
+	needsReboot := false
+	for b := range needsRebootChan {
+		needsReboot = needsReboot || b
+	}
+
+	var finalErr error
+	for err := range errChan {
+		finalErr = errors.Join(finalErr, err)
+	}
+
+	return needsReboot, finalErr
+}
+
+func hasPendingRebootDBParams(rdsClient *rds.Client, dbId string) (bool, error) {
+	describeParams := &rds.DescribeDBInstancesInput{
+		DBInstanceIdentifier: model.NewPointer(dbId),
+	}
+	describeOut, err := rdsClient.DescribeDBInstances(context.Background(), describeParams)
+	if err != nil {
+		return false, fmt.Errorf("error describing DB instance %q: %w", dbId, err)
+	}
+
+	if len(describeOut.DBInstances) < 1 {
+		return false, fmt.Errorf("describe instances returned no instances")
+	}
+
+	for _, group := range describeOut.DBInstances[0].DBParameterGroups {
+		if group.ParameterApplyStatus == nil {
+			return false, fmt.Errorf("parameter group has no ParameterApplyStatus")
+		}
+
+		if *group.ParameterApplyStatus == "pending-reboot" {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (t *Terraform) RebootDBInstances(extAgent *ssh.ExtAgent) error {
+	// Build the RDS client
+	cfg, err := t.GetAWSConfig()
+	if err != nil {
+		return fmt.Errorf("failed to get AWS config: %w", err)
+	}
+	rdsClient := rds.NewFromConfig(cfg)
+
+	// Reboot each DB instance in parallel
+	errChan := make(chan error, len(t.output.DBCluster.Instances))
```

Comment:
Do you ever get into the situation where only one of the dbs need to be rebooted? Or will it always require them all to be rebooted?

Reply:
When we apply DB parameters, we apply them to every instance, so I don't think there's any scenario like that. Even if there was one, I think it's worth rebooting all instances: it won't make the process longer, since we're doing it concurrently, and it makes the logic slightly simpler.